### PR TITLE
Denoising Diffusion Probabilistic Models for SpeechBrain

### PR DESCRIPTION
**Pull request by flexthink**
This PR contains a basic implementation of Denoising Diffusion Probabilistic Models (DDPM) for SpeechBrain
https://arxiv.org/pdf/2006.11239.pdf

An example is provided to generate MEL-spectrograms using the AudioMNIST dataset
It also contains an implementation of DiffWave by @BenoitWang

### CODE REVIEW

**Comment by mravanelli** [comment]
Thank you @flexthink  for this amazing PR! Let me share my first batch of comments:

1. The current data preparation part (based on HF datasets) is a bit unconventional. To be compliant with the other recipes, we need an external file (`data_preparation.py`) that downloads the data and creates json or csv files (that should be saved in the `save folder`). Users also expect to have a variable called `data_folder` that controls where the dataset is downloaded. In my case,  by default, the dataset is downloaded in my home, which is a poisoning folder in HPC clusters.
2. It looks like the current code requires `librosa` and `soundfile` as additional dependencies. Do we really need it? Also, all the additional dependencies should be reported in a file called `extra-requirements.txt`.
3. When I run on my laptop everything seems to run. However, when I try it on the Mila cluster I got:
```
Traceback (most recent call last):
  File "/network/scratch/r/ravanelm/speechbrain_diffusion/recipes/AudioMNIST/diffusion/train.py", line 780, in <module>
    diffusion_brain.fit(
  File "/network/scratch/r/ravanelm/speechbrain_diffusion/speechbrain/core.py", line 1144, in fit
    self.on_fit_start()
  File "/network/scratch/r/ravanelm/speechbrain_diffusion/speechbrain/core.py", line 795, in on_fit_start
    self.init_optimizers()
  File "/network/scratch/r/ravanelm/speechbrain_diffusion/speechbrain/core.py", line 815, in init_optimizers
    self.optimizer = self.opt_class(self.modules.parameters())
TypeError: __init__() got an unexpected keyword argument 'capturable'
```

Any idea?

**Comment by mravanelli** [comment]
The latter issue disappears with a fresh environment (with pytorch 1.13), so it might be an issue I had on my side.

**Comment by flexthink** [comment]
@mravanelli: The PR is passing the checks and can be reviewed

**Comment by mravanelli** [comment]
Thank you @flexthink and @BenoitWang for submitting this PR! The additions of diffusion and latent diffusion are valuable contributions to our toolkit. After inspecting the code, I have some the following first batch of comments:

1. **Docstring Examples:** While many of the new functions and classes already have docstring examples, it would be beneficial to include more examples. Specifically, I recommend adding examples for the following classes or functions in their respective files:

   - `recipes/AudioMNIST/diffusion/custom_model.py`: `DoneDetector`
   - `speechbrain/lobes/models/DiffWave.py`: `DiffusionEmbedding`, `SpectrogramUpsampler`, `ResidualBlock`, `DiffWave`, `DiffWaveDiffusion` @BenoitWang 
   - `speechbrain/nnet/autoencoder.py`: `Autoencoder`, `VariationalAutoencoder`, `NormalizingAutoencoder`
   - `speechbrain/nnet/diffusion.py`: `Diffuser`, `DenoisingDiffusion`, `LatentDiffusion`, `LengthMaskedGaussianNoise`
   - `speechbrain/nnet/unet.py`: `AttentionPool2d`, `Upsample`, `Downsample`, `ResBlock`, `AttentionBlock`, `QKVAttention`, `UNetModel`, `EncoderUNetModel`, `EmbeddingProjection`, `DecoderUNetModel`, `DownsamplingPadding`, `UNetNormalizingAutoencoder`

2. **Backward Compatibility:** As far as I cab see, there are no backward incompatible changes. However, please @BenoitWang  double-check the use of `.squeeze()` in `speechbrain/lobes/models/FastSpeech2.py` to ensure that replacing it with `.squeeze(-1)` is appropriate.

3. **audiomnist-prepare-data:** To simplify things and avoid pushing a specific folder with relatively large files, I suggest uploading the processed dataset to our Dropbox. This is permitted by the AudioMnist license, but we must include their copyright notice in our modified version of the dataset (https://github.com/soerenab/AudioMNIST/blob/master/LICENSE). Uploading the processed version to Dropbox will also make the preparation script easier, as it can automatically download the processed dataset.

4. **recipes/AudioMNIST/diffusion/README.md:** More details should be added to the README, such as explaining what a diffusion model is and highlighting the differences between conditioned and unconditioned generation, as well as latent and simple diffusion. Additionally, we need to add links to the output folders for each released recipe. Please share the output folder with me so that I can upload them to the official speechbrain dropbox folder and update the links accordingly.

5. **recipes/AudioMNIST/diffusion/custom_model.py:** This function seems to be quite general and could potentially be used beyond AudioMnist. I suggest moving it to `speechbrain.nnet` and creating a library called `utils.py` for classes like this.

6. **recipes/AudioMNIST/diffusion/hparams/train_conditioned.yaml:** There seems to be some confusion here. This recipe should not inject any conditioning embeddings, but it looks like both speaker and digit embeddings are injected, with `speaker_conditioned: True`. Is this a mistake?

7. **speechbrain/nnet/autoencoder.py:** Since there are multiple encoders here, it might be more appropriate to rename the file to `autoencoders.py`.

8. **speechbrain/nnet/losses.py:** The newly introduced `mask_shape="targets"` is not described in the docstring. Please add a description for this parameter.

9. **speechbrain/pretrained/interfaces.py:** There is a typo: `HPARAMS_NEEDED = ["difussion"]` should be `HPARAMS_NEEDED = ["diffusion"]`. Additionally, we need to verify whether we already have this model on Hugging Face. If not, we should add it.

10. **tests/consistency/test_recipe.py:** What is the reason for adding `/pretrained_models/` in `test_recipes`?

11. **tests/samples/AudioMNIST/:** These samples are only needed for running the recipe tests, so we don't need real data here, just fake data. To keep the repo small, could we reuse the data already available in `tests/samples` and only push a CSV file with "fake" labels?

12. **speechbrain/dataio/dataset.py:** I believe `apply_overfit_test` should not be here. My suggestion is to add it to the `train.py` of the recipe(s) that use it.

13. **HF Repo for DiffWave:** We need to create the Hugging Face repository for DiffWave and put the output folder on the speechbrain dropbox. @BenoitWang , please follow up with me for that.

14. **Writing a Tutorial:** I propose writing a tutorial on diffusion, as this is independent of the current PR and will be a valuable addition to our documentation.

Feel free to address these comments. Once done, I will run an actually training experiment.

File: speechbrain/lobes/models/FastSpeech2.py
Comment (BenoitWang) [inline speechbrain/lobes/models/FastSpeech2.py:1]: I just re-ran the fastspeech training to check the shapes and I found squeeze(-1) is correct. Everything works fine, thanks @flexthink, a good catch!

**Review by mravanelli — APPROVED** [review]

**Comment by mravanelli** [comment]
I did extensive tests. I ran the full diffusion recipes. Everything looks good to me now. The quality of the digit generated with diffusion is pretty high, considering that AudioMNIST is a small dataset. Only the quality of latent diffusion is quite low, but I think we can improve it in another PR. Thank you  @flexthink and @BenoitWang for this great work!

**Pull request merged**